apis/r/R/SOMADataFrame.R

```diff
@@ -0,0 +1,157 @@
+          name = field_name,
+          type = field_type,
+          nullable = field$nullable,
+          ncells = if (field_type == "ASCII") NA_integer_ else 1L,
+          filter_list = tiledb::tiledb_filter_list(c(
+            tiledb_zstd_filter()
+          ))
+        )
+      }
+
+      # array schema
+      tdb_schema <- tiledb::tiledb_array_schema(
+        domain = tiledb::tiledb_domain(tdb_dim),
+        attrs = tdb_attrs,
+        sparse = TRUE,
+        cell_order = "ROW_MAJOR",
+        tile_order = "ROW_MAJOR",
+        capacity = 100000,
+        # TODO: should be configurable via a global option
+        allows_dups = FALSE,
+        offsets_filter_list = tiledb::tiledb_filter_list(c(
+          tiledb::tiledb_filter("DOUBLE_DELTA"),
+          tiledb::tiledb_filter("BIT_WIDTH_REDUCTION"),
+          tiledb_zstd_filter()
+        ))
+      )
+
+      # create array
+      tiledb::tiledb_array_create(uri = self$uri, schema = tdb_schema)
+    },
+
+    #' @description Write
+    #'
+    #' @param values An [`arrow::Table`] containing all columns, including
+    #' the `soma_rowid` index column. The schema for `values` must match the
+    #' schema for the `SOMADataFrame`.
+    #'
+    write = function(values) {
+      stopifnot(
+        "'values' must be an Arrow Table" =
+          is_arrow_table(values),
+        "'values' must contain a 'soma_rowid' column name" =
+          "soma_rowid" %in% colnames(values)
+      )
+
+      df <- as.data.frame(values)[c(self$dimnames(), self$attrnames())]
+      # coerce from integer to integer64 in order to match the dimension type
+      df$soma_rowid <- bit64::as.integer64(df$soma_rowid)
+
+      on.exit(private$close())
+      private$open("WRITE")
+      arr <- self$object
+      arr[] <- df
+    },
+
+    #' @description Read
+    #' Read a user-defined subset of data, addressed by the dataframe indexing
+    #' column, and optionally filtered.
+    #' @param ids Integer indices specifying the rows to read.
+    #' @param column_names Character vector of column names to return.
+    #' @param value_filter A string containing a logical expression that is used
+    #' to filter the returned values. See [`tiledb::parse_query_condition`] for
+    #' more information.
+    #' @param result_order Order of read results. This can be one of either
+    #' `"ROW_MAJOR, `"COL_MAJOR"`, `"GLOBAL_ORDER"`, or `"UNORDERED"`.
+    #' @return An [`arrow::Table`].
+    read = function(
+      ids = NULL,
+      column_names = NULL,
+      value_filter = NULL,
+      result_order = "UNORDERED"
+    ) {
+      on.exit(private$close())
+      private$open("READ")
+
+      arr <- self$object
+
+      # soma_rowid should not be included in the results
+      tiledb::extended(arr) <- FALSE
+
+      # select columns
+      tiledb::attrs(arr) <- column_names %||% character()
+
+      # select ranges
+      if (!is.null(ids)) {
+        tiledb::selected_ranges(arr) <- list(soma_rowid = cbind(ids, ids))
```

Comment:
does this 'stick' in place after the `read` has returned?

Reply:
bigger picture, `tiledb::selected_ranges` in the `tiledb-r` API is stateful while `read` in the `tiledb-soma` API is stateless

Reply:
I make a copy of the array pointer (`arr <- self$object`) to keep it stateless, e.g.:

```r
library(tiledbsoma)
uri <- file.path(tempdir(), "soma-dataframe")
asch <- arrow::schema(
  foo = arrow::int32(),
  bar = arrow::float64(),
  baz = arrow::string()
)

sdf <- SOMADataFrame$new(uri)
sdf$create(asch)

tbl <- arrow::arrow_table(
  soma_rowid = 0L:9L, foo = 1L:10L, bar = 1.1:10.1, baz = letters[1:10]
)

sdf$read(value_filter = "foo < 5")
# Table
# 4 rows x 3 columns
# $foo <int32>
# $bar <double>
# $baz <string>

sdf$read()
# Table
# 10 rows x 3 columns
# $foo <int32>
# $bar <double>
# $baz <string>
```

Reply:
nice, totally missed that!!